integration_test/cases/repo.exs

```diff
@@ -999,4 +999,134 @@ defmodule Ecto.Integration.RepoTest do
+    {:ok, inserted} = TestRepo.upsert(%Post{title: "first"})
+    {:ok, _updated} = TestRepo.upsert(%Post{id: inserted.id, title: "second"})
+
+    post = TestRepo.one!(Post)
+    assert post.title == "second"
+  end
+
+  @tag :upsert
+  test "upsert with struct uses given columns for conflict" do
+    uuid = Ecto.UUID.generate
+
+    {:ok, _} = TestRepo.upsert(%Post{title: "first", uuid: uuid}, conflict_target: [:uuid])
+    {:ok, _} = TestRepo.upsert(%Post{title: "second", uuid: uuid}, conflict_target: [:uuid])
+    post = TestRepo.one!(Post)
+    assert post.title == "second"
+    {:ok, _} = TestRepo.upsert(%Post{title: "third", uuid: uuid}, conflict_target: :uuid)
+    post = TestRepo.one!(Post)
+    assert post.title == "third"
+  end
+
+  @tag :upsert
+  test "upsert with struct using composite primary key" do
+    {:ok, _} = TestRepo.upsert(%CompositePk{a: 1, b: 2, name: "first"})
+    {:ok, _} = TestRepo.upsert(%CompositePk{a: 1, b: 2, name: "second"})
+
+    record = TestRepo.one!(CompositePk)
+    assert record.name == "second"
+  end
+
+  @tag :upsert
+  test "upsert with changeset uses primary key if given" do
+    insert =
+      Ecto.Changeset.cast(%Post{}, %{"title" => "first"}, ~w(title))
+      |> Ecto.Changeset.unique_constraint(:uuid)
+
+    {:ok, inserted} = TestRepo.upsert(insert)
+
+    # Ignore the UUID unique constraint when the primary key has been set
+    changes = %{"id" => inserted.id, "title" => "second"}
+    update =
+      Ecto.Changeset.cast(%Post{}, changes, ~w(id title))
+      |> Ecto.Changeset.unique_constraint(:uuid)
+    {:ok, _} = TestRepo.upsert(update)
+
+    post = TestRepo.one!(Post)
+    assert post.title == "second"
+  end
+
+  @tag :upsert
+  test "upsert with changeset override conflict target columns" do
+    insert =
+      Ecto.Changeset.cast(%Post{}, %{"title" => "first"}, ~w(title))
+      |> Ecto.Changeset.unique_constraint(:uuid)
+    {:ok, inserted} = TestRepo.upsert(insert)
+
+    # Even though the ID has been explicitly set, use the UUID column
+    changes = %{"id" => inserted.id + 1, "uuid" => inserted.uuid, "title" => "second"}
+    update =
+      Ecto.Changeset.cast(%Post{}, changes, ~w(id uuid title))
+      |> Ecto.Changeset.unique_constraint(:uuid)
+
+    {:ok, _} = TestRepo.upsert(update, conflict_target: [:uuid])
+
+    post = TestRepo.one!(Post)
+    assert post.id == inserted.id + 1
+    assert post.title == "second"
+  end
+
+  @tag :upsert
+  test "upsert on conflict ignore" do
+    {:ok, inserted} = TestRepo.upsert(%Ecto.Integration.Post{title: "first", uuid: "6fa459ea-ee8a-3ca4-894e-db77e160355e"},
+      on_conflict: :nothing)
+
+    assert inserted.id > 0
+
+    {:ok, not_inserted} = TestRepo.upsert(%Ecto.Integration.Post{title: "first", uuid: "6fa459ea-ee8a-3ca4-894e-db77e160355e"},
+      on_conflict: :nothing, update: [:id]) # Update field specified for mysql resulting in ON DUPLICATE KEY `id` = `id`
+    assert not_inserted.id in [0, nil]
```

Comment:
In this case, is it possible for postgres/mysql to know the insert did not change anything to the table? Because if we can know that, then we can make it so none of them change anything at all (i.e. both are nil).